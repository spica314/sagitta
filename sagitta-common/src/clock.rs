#[derive(Debug, Clone)]
pub struct Clock {
    pub fixed_time: Option<std::time::SystemTime>,
}

impl Clock {
    pub fn new() -> Self {
        Self { fixed_time: None }
    }

    pub fn new_with_fixed_time(fixed_time: std::time::SystemTime) -> Self {
        Self {
            fixed_time: Some(fixed_time),
        }
    }

    pub fn now(&self) -> std::time::SystemTime {
        self.fixed_time.unwrap_or_else(std::time::SystemTime::now)
    }

    pub fn is_fixed(&self) -> bool {
        self.fixed_time.is_some()
    }
}

impl Default for Clock {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_clock() {
        let clock = Clock::default();
        let now = clock.now();
        assert!(now.elapsed().unwrap().as_secs() < 1);
    }

    #[test]
    fn test_clock_with_fixed_time() {
        let fixed_time = std::time::SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(42);
        let clock = Clock::new_with_fixed_time(fixed_time);
        let now = clock.now();
        assert_eq!(now, fixed_time);
    }
}
