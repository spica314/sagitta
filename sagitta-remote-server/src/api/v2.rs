pub mod commit;
pub mod create_workspace;
pub mod get_attr;
pub mod get_file_blob_id;
pub mod get_workspace_id_from_name;
pub mod get_workspaces;
pub mod read_blob;
pub mod read_dir;
pub mod sync_files_with_workspace;
pub mod write_blob;
