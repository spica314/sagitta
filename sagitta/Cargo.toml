[package]
name = "sagitta"
version = "0.1.0"
edition = "2021"

[dependencies]
clap = { workspace = true }
env_logger = { workspace = true }
fuser = { workspace = true }
libc = { workspace = true }
log = { workspace = true }
serde_cbor = { workspace = true }
tokio = { workspace = true }
ureq = { workspace = true }

sagitta-common = { workspace = true }
sagitta-local-api-client = { workspace = true }
sagitta-local-api-schema = { workspace = true }
sagitta-local-server = { workspace = true }
sagitta-local-system-workspace = { workspace = true }
sagitta-remote-api-client = { workspace = true }
sagitta-remote-api-schema = { workspace = true }
