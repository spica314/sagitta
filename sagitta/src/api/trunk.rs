pub mod get_head;
