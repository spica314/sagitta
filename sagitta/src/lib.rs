pub mod api_client;
pub mod args;
pub mod fs;
