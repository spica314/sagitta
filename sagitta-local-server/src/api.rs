use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;

use crate::api_state::ApiState;

use self::v1::sync::v1_sync;

pub mod v1;

pub struct ServerConfig {
    pub clock: Clock,
    pub port: u16,
    pub local_system_workspace_base_path: PathBuf,
    pub remote_api_base_url: String,
}

pub async fn run_local_api_server(config: ServerConfig) {
    let state = ApiState::new(
        config.clock.clone(),
        config.local_system_workspace_base_path.clone(),
        &config.remote_api_base_url,
    )
    .await;

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(v1_sync)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
