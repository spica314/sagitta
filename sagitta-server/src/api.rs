use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;
use sagitta_objects::{SagittaTreeObject, SagittaTreeObjectFile};

use crate::state::ApiState;

use self::{
    blob::read::blob_read,
    trunk::get_head::trunk_get_head,
    workspace::{create::workspace_create, get_head::workspace_get_head, list::workspace_list},
};

pub mod blob;
pub mod trunk;
pub mod workspace;

pub struct ServerConfig {
    pub base_path: PathBuf,
    pub surreal_uri: String,
    pub is_main: bool,
    pub clock: Clock,
    pub port: u16,
}

pub async fn run_server(config: ServerConfig) {
    let state = ApiState::new(
        config.base_path.clone(),
        config.clock.clone(),
        &config.surreal_uri,
        config.is_main,
    )
    .await;

    // root (dir1)
    // - hello.txt (file1)
    // - hello_dir (dir2)
    //     - hello2.txt (file2)

    let file1_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!\n".as_slice())
        .unwrap();
    let file1 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file1_blob_id,
        size: 14,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file1)
        .unwrap();

    let file2_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!!\n".as_slice())
        .unwrap();
    let file2 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file2_blob_id,
        size: 15,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file2)
        .unwrap();

    let tree2 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![("hello2.txt".to_string(), file2_id)],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree2)
        .unwrap();

    let tree1 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![
            ("hello.txt".to_string(), file1_id),
            ("hello_dir".to_string(), tree2_id),
        ],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree1)
        .unwrap();

    let commit = sagitta_objects::SagittaCommitObject {
        tree_id: tree1_id,
        parent_commit_id: None,
        message: "Initial commit".to_string(),
    };
    let commit_id = state
        .remote_system_workspace_manager
        .save_commit(None, &commit)
        .unwrap();
    state
        .remote_system_workspace_manager
        .set_trunk_head(&commit_id)
        .await
        .unwrap();

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(blob_read)
            .service(trunk_get_head)
            .service(workspace_create)
            .service(workspace_list)
            .service(workspace_get_head)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
