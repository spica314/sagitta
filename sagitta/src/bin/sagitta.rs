use std::{path::PathBuf, str::FromStr};

use clap::Parser;
use sagitta::args::Args;
use sagitta::fs::{run_fs, SagittaConfig};
use sagitta_common::clock::Clock;
use sagitta_local_api_schema::v1::sync::V1SyncRequest;
use sagitta_local_server::api::ServerConfig;
use sagitta_remote_api_schema::v2::create_workspace::V2CreateWorkspaceRequest;
use sagitta_remote_api_schema::v2::get_workspaces::{
    V2GetWorkspacesRequest, V2GetWorkspacesResponse,
};

#[tokio::main]
async fn main() {
    env_logger::init();

    let args = Args::parse();
    eprintln!("args = {:?}", args);

    if let Some(mount) = &args.mount {
        let uid = unsafe { libc::getuid() };
        let gid = unsafe { libc::getgid() };
        let config = SagittaConfig {
            base_url: "http://localhost:8512".to_string(),
            mountpoint: mount.to_string(),
            uid,
            gid,
            clock: Clock::new(),
            local_system_workspace_base_path: PathBuf::from_str("./sagitta-test-system").unwrap(),
            debug_sleep_duration: None,
        };
        let api_config = ServerConfig {
            clock: config.clock.clone(),
            port: 8513,
            local_system_workspace_base_path: config.local_system_workspace_base_path.clone(),
            remote_api_base_url: config.base_url.clone(),
        };
        tokio::spawn(async move {
            sagitta_local_server::api::run_local_api_server(api_config).await;
        });
        run_fs(config);
    } else {
        let api_client =
            sagitta_remote_api_client::SagittaApiClient::new("http://localhost:8512".to_string());
        let local_api_client = sagitta_local_api_client::SagittaLocalApiClient::new(
            "http://localhost:8513".to_string(),
        );
        let command = args.subcommand.unwrap();
        match command {
            sagitta::args::Commands::Workspace { subcommand } => match subcommand.unwrap() {
                sagitta::args::WorkspaceSubcommands::Create { name } => {
                    let list = api_client
                        .v2_get_workspaces(V2GetWorkspacesRequest {})
                        .unwrap();
                    match list {
                        V2GetWorkspacesResponse::Ok { items } => {
                            if items.iter().any(|workspace| workspace.name == name) {
                                eprintln!("Workspace {} already exists", name);
                                return;
                            }
                        }
                        _ => {
                            eprintln!("Failed to get workspaces");
                            return;
                        }
                    }
                    api_client
                        .v2_create_workspace(V2CreateWorkspaceRequest { name })
                        .unwrap();
                }
                sagitta::args::WorkspaceSubcommands::List => {
                    let list = api_client
                        .v2_get_workspaces(V2GetWorkspacesRequest {})
                        .unwrap();
                    match list {
                        V2GetWorkspacesResponse::Ok { items } => {
                            for workspace in items {
                                println!("{}", workspace.name);
                            }
                        }
                        _ => {
                            eprintln!("Failed to get workspaces");
                        }
                    }
                }
            },
            sagitta::args::Commands::Sync { workspace_id } => {
                local_api_client
                    .v1_sync(V1SyncRequest {
                        workspace_id: workspace_id.clone(),
                    })
                    .unwrap();
            }
        }
    }
}
