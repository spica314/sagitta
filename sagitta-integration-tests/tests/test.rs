use std::{process::Command, time::SystemTime};

use sagitta::{
    api_client::SagittaApiClient,
    fs::{run_fs, SagittaConfig},
};
use sagitta_common::clock::Clock;
use sagitta_local_system_workspace::LocalSystemWorkspaceManager;
use sagitta_objects::SagittaTreeObject;
use sagitta_server::api::ServerConfig;
use serial_test::serial;
use tempfile::tempdir;
use tokio::runtime::Builder;

#[test]
#[serial]
fn test_1() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let port = 8081;
    let config = ServerConfig {
        base_path: tempdir.as_ref().to_path_buf(),
        clock: Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let client = SagittaApiClient::new(format!("http://localhost:{}", port));

    let head_id_res = client.trunk_get_head().unwrap();
    let head_id = head_id_res.id;
    insta::assert_debug_snapshot!(head_id);

    let commit = client.blob_read_as_commit_object(&head_id).unwrap();
    insta::assert_debug_snapshot!(commit);

    let dir = client.blob_read_as_tree_object(&commit.tree_id).unwrap();
    insta::assert_debug_snapshot!(dir);

    let SagittaTreeObject::Dir(dir) = dir else {
        panic!()
    };
    let mut xs = vec![];
    for item in &dir.items {
        let file = client.blob_read_as_tree_object(&item.1).unwrap();
        let SagittaTreeObject::File(file) = file else {
            continue;
        };

        let blob_res = client.blob_read(&file.blob_id).unwrap();
        let blob = std::str::from_utf8(blob_res.blob.as_slice()).unwrap();
        insta::assert_debug_snapshot!(blob);

        xs.push((item.0.clone(), blob.to_string()));
    }
    insta::assert_debug_snapshot!(xs);
}

#[test]
#[serial]
fn test_2() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir1 = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let port = 8082;
    let config = ServerConfig {
        base_path: tempdir1.as_ref().to_path_buf(),
        clock: Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let local_system_workspace_base_path = tempdir().unwrap().as_ref().to_path_buf();

    let tempdir2 = tempdir().unwrap();
    let tempdir2_str = tempdir2.as_ref().to_str().unwrap().to_string();
    let uid = unsafe { libc::getuid() };
    let gid = unsafe { libc::getgid() };
    std::thread::spawn(move || {
        let config = SagittaConfig {
            base_url: format!("http://localhost:{}", port),
            mountpoint: tempdir2_str,
            uid,
            gid,
            clock: Clock::new_with_fixed_time(fixed_system_time),
            local_system_workspace_base_path,
            debug_sleep_duration: None,
        };
        run_fs(config);
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let out0 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(tempdir2.path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out0);

    let mut out1_dir = tempdir2.path().to_path_buf();
    out1_dir.push("trunk");
    let out1 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(out1_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out1);

    let mut out2_dir = tempdir2.path().to_path_buf();
    out2_dir.push("trunk");
    let out2 = Command::new("cat")
        .arg("hello.txt")
        .current_dir(out2_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out2);

    let mut out3_dir = tempdir2.path().to_path_buf();
    out3_dir.push("trunk");
    out3_dir.push("hello_dir");
    let out3 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(&out3_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out3);

    let out4 = Command::new("cat")
        .arg("hello2.txt")
        .current_dir(out3_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out4);
}

#[test]
#[serial]
fn test_3() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir1 = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let port = 8083;
    let config = ServerConfig {
        base_path: tempdir1.as_ref().to_path_buf(),
        clock: Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let client = SagittaApiClient::new(format!("http://localhost:{}", port));
    client.workspace_create("workspace1").unwrap();
    let workspaces = client.workspace_list().unwrap();
    insta::assert_debug_snapshot!(workspaces);
}

#[test]
#[serial]
fn test_4() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir1 = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    // port 8084 is used by GitHub Actions
    let port = 8085;
    let config = ServerConfig {
        base_path: tempdir1.as_ref().to_path_buf(),
        clock: Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let client = SagittaApiClient::new(format!("http://localhost:{}", port));
    client.workspace_create("workspace1").unwrap();

    let local_system_workspace_base_path = tempdir().unwrap().as_ref().to_path_buf();

    let tempdir2 = tempdir().unwrap();
    let tempdir2_str = tempdir2.as_ref().to_str().unwrap().to_string();
    {
        let local_system_workspace_base_path = local_system_workspace_base_path.clone();
        let uid = unsafe { libc::getuid() };
        let gid = unsafe { libc::getgid() };
        std::thread::spawn(move || {
            let config = SagittaConfig {
                base_url: format!("http://localhost:{}", port),
                mountpoint: tempdir2_str,
                uid,
                gid,
                clock: Clock::new_with_fixed_time(fixed_system_time),
                local_system_workspace_base_path: local_system_workspace_base_path.clone(),
                debug_sleep_duration: None,
            };
            run_fs(config);
        });
        std::thread::sleep(std::time::Duration::from_secs(1));
    }

    let local_system_workspace_manager =
        LocalSystemWorkspaceManager::new(local_system_workspace_base_path.clone());

    {
        let bytes = b"Hello, copy on write!";
        std::fs::write(tempdir2.path().join("workspace1").join("cow.txt"), bytes).unwrap();
    }

    local_system_workspace_manager
        .create_cow_file(
            "workspace1",
            &["cow_dir".to_string(), "cow.txt".to_string()],
            b"Hello, copy on write! (dir)",
        )
        .unwrap();

    let out0 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(tempdir2.path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out0);

    let workspace_path = tempdir2.path().join("workspace1");
    let out1 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(workspace_path.as_path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out1);

    let out2 = Command::new("cat")
        .arg("cow.txt")
        .current_dir(workspace_path.as_path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out2);

    let workspace_path = tempdir2.path().join("workspace1").join("cow_dir");
    let out3 = Command::new("ls")
        .arg("-lAUgG")
        .current_dir(workspace_path.as_path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out3);

    let out4 = Command::new("cat")
        .arg("cow.txt")
        .current_dir(workspace_path.as_path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out4);

    let path_out5 = tempdir2.path().join("workspace1");
    Command::new("bash")
        .arg("-c")
        .arg("echo 'Hello, copy on write! (overwrite)' > cow.txt")
        .current_dir(&path_out5)
        .output()
        .expect("failed to execute process");
    let out5 = Command::new("cat")
        .arg("cow.txt")
        .current_dir(&path_out5)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out5);

    let path_out6 = tempdir2.path().join("trunk");
    let out6 = Command::new("bash")
        .arg("-c")
        .arg("echo 'Hello (must fail)' > fail.txt")
        .current_dir(&path_out6)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out6);

    let path_out7 = tempdir2.path().join("workspace1");
    Command::new("bash")
        .arg("-c")
        .arg("echo 'Hello, copy on write! (truncate_test)' > truncate.txt")
        .current_dir(&path_out5)
        .output()
        .expect("failed to execute process");
    Command::new("bash")
        .arg("-c")
        .arg("echo 'Hello, copy on write! (truncate)' > truncate.txt")
        .current_dir(&path_out5)
        .output()
        .expect("failed to execute process");
    let out7 = Command::new("cat")
        .arg("truncate.txt")
        .current_dir(&path_out7)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out7);
}
