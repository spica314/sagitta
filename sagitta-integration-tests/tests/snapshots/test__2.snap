---
source: sagitta-integration-tests/tests/test.rs
expression: out1
---
Output {
    status: ExitStatus(
        unix_wait_status(
            0,
        ),
    ),
    stdout: "total 0\n-rw-r--r-- 1 root root 14 Dec 22  2009 hello.txt\ndrwxr-xr-x 2 root root  0 Dec 22  2009 hello_dir\n",
    stderr: "",
}
